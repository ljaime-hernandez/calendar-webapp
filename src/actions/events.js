import Swal from "sweetalert2";
import { fetchWithToken } from "../helpers/fetch";
import { fixEvents } from "../helpers/fixEvents";
import { types } from "../types/types";

export const eventStartAddNew = (event) => {

    return async(dispatch, getState) => {
        
        // we use the getState to retrieve the uid and name in the store dispatched to our authReducer when
        // the user logged on its account
        const { uid, name } = getState().auth;

        try {

        // the fetchWithToken function is used as, for this request to work, an authorized user must be logged in, 
        // if the user is authenticated and the token is still valid, then the event data will go through some checks
        // on the backend middlewares, to finally be registered in our mongo database. The first argument will be the 
        // endpoint which is 'events' and the third argument must be a POST
            const res = await fetchWithToken('events', event, 'POST');
            const body = await res.json();

                // if the response is correct, then we will add the uid and name of the user for it to be used as reference
                // on the eventModal, which will help us register the event using this user as reference for whenever we want
                // to retrieve all events registered on it
                if(body.ok){
                    event.id = body.event.id;
                    event.user = {
                        _id: uid,
                        name: name
                    };
                    // the event information will be dispatched to an array of events which we will be able to track from the store
                    // if necessary
                    dispatch(eventAddNew(event));
                };
            
        } catch (error) {
            console.log(error);
        };
    };
};

export const eventStartLoading = () => {

    return async(dispatch) => {

        try {
            
            // The fetchWithToken will do a GET request by default and will use the endpoint to receive all the events
            // received from the calendar backend for them to later be displayed in the users calendar display.
            const resp = await fetchWithToken('events');
            const body = await resp.json();
            // The start and end dates on the events might not work properly, so we need a helper for us to change the
            // dates registered in the mongo database and modify them with moment.js for us to use them properly
            const events = fixEvents(body.events);
            
            dispatch(eventLoaded(events));

        } catch (error) {
            console.log(error);
        };
    };
} ;

export const eventStartUpdate = (event) => {

    return async(dispatch) => {

        try {
            
            // this fetch will use the PUT as method to update the event by using its id and the events endpoint as
            // reference, if the event is found then the calendar backend will use a mongo method for the event to be
            // updated
            const resp = await fetchWithToken(`events/${event.id}`, event, 'PUT');
            const body = await resp.json();

            if(body.ok) {
                dispatch(eventUpdated(event));
                dispatch(eventStartLoading());
            } else {
                // the error message displayed in the Sweetalert popup is retrieved from the calendar backend, 
                // which will be attached to the body response 
                Swal.fire('Error', body.msg, 'error');
            };

        } catch (error) {
            console.log(error);
        };
    };
};

export const eventStartDelete = () => {

    return async(dispatch, getState) => {

        // we use the getState to retrieve the id in the store dispatched to our calendarReducer when
        // the user clicks on an specific event
        const { id } = getState().calendar.activeEvent; 

        try {
            
            // Similar to the eventStartUpdate, this fetch will use the DELETE as method to delete the event by using 
            // its id and the events endpoint as reference, if the event is found then the calendar backend will use 
            // a mongo method for the event to be deleted and no additional event data needs to be sent as the deletion
            // only requires the id for it
            const resp = await fetchWithToken(`events/${id}`, {}, 'DELETE');
            const body = await resp.json();

            if(body.ok) {
                dispatch(eventDeleted());
                dispatch(eventStartLoading());
                dispatch(eventClearActive());
            } else {
                // the error message displayed in the Sweetalert popup is retrieved from the calendar backend, 
                // which will be attached to the body response 
                Swal.fire('Error', body.msg, 'error');
            };
        } catch (error) {
            console.log(error);
        };
    };
};

export const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
});

export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
});

export const eventClearActive = () => ({
    type: types.eventClearActive
});

const eventUpdated = (event) => ({
    type: types.eventUpdated,
    payload: event
})

const eventDeleted = () => ({
    type: types.eventDeleted
})

const eventLoaded = (events) => ({
    type: types.eventLoaded,
    payload: events
})

export const eventLogout = () => ({
    type: types.eventLogout
}) 