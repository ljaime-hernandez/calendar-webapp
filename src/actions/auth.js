import Swal from "sweetalert2";
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";
import { eventLogout } from "./events";

export const startLogin = (email, password) => {
    
    return async(dispatch) => {

        // the fetchWithoutToken function is used as the user is not logged in and we are about to create a token
        // for it if the user exists in our database, the first argument will be the endpoint which is the
        // auth and the third argument must be a POST
        const resp = await fetchWithoutToken('auth', {email, password}, 'POST');
        const body = await resp.json();

        // if we receive the correct response, we will use the localStorage to temporarily place our received token,
        // by default the token will last for 2 hours, but it can be renewed with some of the functions created on our
        // calendar, we also place the specific time where the token was placed in the token-init-date portion of the
        // localStorage to know if the token is still valid.
        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            // the dispatch will be sent to the store for it to be registered on the reducer
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));

        } else {
            // the error message displayed in the Sweetalert popup is retrieved from the calendar backend, 
            // which will be attached to the body response 
            Swal.fire('error', body.msg, 'error');
        };
    };
};

// The startRegister function will not only register a new user on the database but it will also 
// automatically log the user in.
export const startRegister = (name, email, password) => {

    return async(dispatch) => {

        // The fetchWithoutToken function is used as the user is not logged in and we are about to create a token
        // for it. The first argument will be the endpoint which is the auth/new and the third argument must be a POST
        const resp = await fetchWithoutToken('auth/new', {name, email, password}, 'POST');
        const body = await resp.json();

        // If we receive the correct response, we will use the localStorage to temporarily place our received token,
        // by default the token will last for 2 hours, but it can be renewed with some of the functions created on our
        // calendar, we also place the specific time where the token was placed in the token-init-date portion of the
        // localStorage to know if the token is still valid.
        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            dispatch(register({
                uid: body.uid,
                name: body.name
            }));

            // the dispatch will be sent to the store for it to be registered on the reducer
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));

        } else {
            // the error message displayed in the Sweetalert popup is retrieved from the calendar backend, 
            // which will be attached to the body response 
            Swal.fire('Error', body.msg, 'error');
        };
    };
};

export const startChecking = () => {

    return async(dispatch) => {

        // the endpoint of the fethWithToken uses the renew portion of it for the user to retrieve an updated token for the 
        // user to continue its operations on the webapp for a longer period of time
        const resp = await fetchWithToken('auth/renew');
        const body = await resp.json();

        // If we receive the correct response, we will use the localStorage to temporarily place our received token,
        // by default the token will last for 2 hours, but it can be renewed with some of the functions created on our
        // calendar, we also place the specific time where the token was placed in the token-init-date portion of the
        // localStorage to know if the token is still valid.
        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            dispatch(checkDone());
            // the dispatch will be sent to the store for it to be registered on the reducer
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        } else {
            dispatch(checkDone());
        };
    };
};

export const startLogout = () => {

    return (dispatch) => {

        // once the user logs out, we will clean every piece of information from the localStorage
        localStorage.clear();
        dispatch(eventLogout());
        dispatch(logout());
    }
}

const login = (user) => ({
    type: types.authLogin,
    payload: user
})

const register = (user) => ({
    type: types.authStartRegister,
    payload: user
})

const checkDone = () => ({type: types.authCheckDone});

const logout = () => ({type: types.authLogout});