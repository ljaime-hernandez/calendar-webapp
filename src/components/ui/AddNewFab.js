import React from 'react'
import { useDispatch } from 'react-redux';
import { uiOpenModal } from '../../actions/ui';

export const AddNewFab = () => {
    const dispatch = useDispatch();

    // when the AddNewFab button is clicked, it will change the modalOpen value status from false to true,
    // opening the modal to allow the user to add a new event into our mongo database
    const handleClickNew = () => {
        dispatch(uiOpenModal());
    }
  
    return (
    <button
        className='btn btn-primary fab'
        onClick={handleClickNew}
    >
        <i className='fas fa-plus'></i>
    </button>
  )
}
