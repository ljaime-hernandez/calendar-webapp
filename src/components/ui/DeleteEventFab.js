import React from 'react'
import { useDispatch } from 'react-redux';
import { eventStartDelete } from '../../actions/events';

export const DeleteEventFab = () => {
    
    const dispatch = useDispatch();
    // The click event on this component will trigger a delete request on our calendar backend for the activeEvent
    // to be deleted, for more information on how this function works please refer to the events action file
    const handleDelete = () => {
        dispatch(eventStartDelete());
    };

  return (
    <button
        className='btn btn-danger fab-danger'
        onClick={handleDelete}    
    >
        <i className='fas fa-trash'></i>
        <span>Delete Event</span>
    </button>
  );
};
