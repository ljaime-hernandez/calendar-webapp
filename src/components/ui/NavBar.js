import React from 'react';
import { Nav, Navbar } from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';
import { startLogout } from '../../actions/auth';

export const NavBar = () => {

  const dispatch = useDispatch();
  // the useSelector will retrieve the user name from the authReducer to display it dynamically on the 
  // navbar
  const { name } = useSelector( state => state.auth );

  const handleLogout = () => {
    // the logout function will erase all information saved in the localStorage and store, for additional
    // information please refer to the auth action file
    dispatch(startLogout());
  }

  return (
    <Navbar collapseOnSelect expand="lg" bg="dark" variant="dark">
        <Navbar.Brand href="#home">{name}</Navbar.Brand>
        <Navbar.Toggle aria-controls="responsive-navbar-nav" />
        <Navbar.Collapse id="responsive-navbar-nav">
            <Nav className="me-auto ">
            <button 
              className='btn btn-outline-danger'
              onClick={handleLogout}>
                <i className="fa-solid fa-arrow-right-from-bracket"></i>
                &nbsp;&nbsp;Logout
            </button>
            </Nav>
        </Navbar.Collapse>
        </Navbar>
  )
}
