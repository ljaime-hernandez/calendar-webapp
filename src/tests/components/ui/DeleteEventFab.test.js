import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { DeleteEventFab } from '../../../components/ui/DeleteEventFab';
import '@testing-library/jest-dom';
import { eventStartDelete } from '../../../actions/events';

jest.mock('../../../actions/events', () => ({
    eventStartDelete: jest.fn()
}))

const middlewares = [ thunk ];
const mockStore = configureStore(middlewares);

const initState = {};
let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider
        store={store}
    >
        <DeleteEventFab/>
    </Provider>
)

describe('Tests on DeleteEventFab component', () => {

    test('should render properly', () => { 

        expect(wrapper).toMatchSnapshot();
    });

    test('should call EventStartDelete on click', () => { 

        wrapper.find('button').prop('onClick')();
        expect(eventStartDelete).toHaveBeenCalled();
    });
})