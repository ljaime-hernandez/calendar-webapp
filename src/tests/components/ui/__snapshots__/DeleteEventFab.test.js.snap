// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on DeleteEventFab component should render properly 1`] = `
<button
  className="btn btn-danger fab-danger"
  onClick={[Function]}
>
  <i
    className="fas fa-trash"
  />
  <span>
    Delete Event
  </span>
</button>
`;
